### DISCUSSION TITLE
How do I make `vim.fn.feedkeys('d', 'n')` enter operator-pending mode?

### DISCUSSION THREAD

**Discussion by nyngwang** [neovim/neovim#24157] [Q&A] [answered]
As title. I have a keymap like this:

```lua
vim.keymap.set('n', 'd',  function () if dapui_open and G.cur_tab(tab_debug) then return end vim.fn.feedkeys('d', 'n') end)
```

I found that `diw` was treated as `idw`, i.e. it entered insert mode and inserted `dw`. How do I fix it?

**Comment by nyngwang** [comment, marked as answer]
It turns out that my other `d`-prefix keymaps, i.e. `d{d,m,c,r,t,k, ...}`, are the causes of the sequence `di` being executed as `id`. To work around this, we can create a mapping for `di` too:

```lua
vim.keymap.set('n', 'di',  function () if dapui_open and G.cur_tab(tab_debug) then return end vim.fn.feedkeys('di', 'n') end)
```

**Reply by nyngwang** [reply]
You still need to read the [answer](https://github.com/neovim/neovim/discussions/24157#discussioncomment-6277451) below to replace `vim.fn.feedkeys` with the `expr` argument.

**Comment by zeertzjq** [comment]
It's better to use an `expr` mapping ~and `vim.schedule()` the `G.curtab()`~ here.

**Reply by nyngwang** [reply]
Thank you. The reason I didn't use `expr` because I thought using it wouldn't be able to call a function, i.e. `G.cur_tab` here. I will try your advice.

For more context (apologies that I omitted it) the `G.cur_tab` is defined like this:

```lua
function M.cur_tab(tab)
  if not tab then tab = -1 end
  return vim.api.nvim_get_current_tabpage() == tab
end
```

**Reply by nyngwang** [reply]
Does putting `vim.schedule()` inside/outside `G.curtab` make a difference?

**Reply by zeertzjq** [reply]
Oh, `vim.schedule()` isn't needed here at all.

**Reply by nyngwang** [reply]
I have tried `expr` and it works well :smile:, but now I have a little problem with the following three `keymap`s that `dap.{terminate,step_back,step_over}` will only be executed **after** the next any key that follows `d{t,k,j}`. Do you know what's the cause? (my `G.EXPR_NOREF...` is `{expr=true, noremap=true, silent=true, nowait=true}`)

```lua
vim.keymap.set('n', 'dt', function () if dapui_open and G.cur_tab(tab_debug) then dap.terminate() return end return 'dt' end, G.EXPR_NOREF_NOERR_TRUNC)
vim.keymap.set('n', 'dk', function () if dapui_open and G.cur_tab(tab_debug) then dap.step_back() return end return 'dk' end, G.EXPR_NOREF_NOERR_TRUNC)
vim.keymap.set('n', 'dj', function () if dapui_open and G.cur_tab(tab_debug) then dap.step_over() return end return 'dj' end, G.EXPR_NOREF_NOERR_TRUNC)
```

**Reply by nyngwang** [reply]
Nevermind, I found the solution by

```diff
- dap.terminate() return end -- ...
+ dap.terminate() return '<NOP>' end --..
```

**Reply by drewlwhitney** [reply]
I know this is so old but thank you so much for commenting this. I was able to remove my use of `nvim_feedkeys()`. I had *no idea* what `expr = true` did.

For anyone that reads this, doing `vim.keymap.set(..., {expr = true})` makes the keybind call a function to determine what the right-hand-side is.

So you can do something like

```
vim.keymap.set(
    "n",
    "some binding",
    function()
        return math.random(100) .. "j" -- string concatenation, so outputs something like "52j"
    end,
    { expr = true }
)
```

and every time you trigger `some binding`, you will move down a random number of lines
